source/crud/bulk-write.md

```diff
@@ -0,0 +1,787 @@
+| `n` | The number of documents that were inserted. | The number of documents that matched the filter. | The number of documents that were deleted. |
+| `nModified` | Not present. | The number of documents that were modified. | Not present. |
+| `upsertedId` | Not present. | The `_id` value for the upserted document. Only present if an upsert took place. | Not present. |
+
+Note that the responses do not contain information about the type of operation that was performed.
+Drivers may need to maintain the user's list of write models to infer which type of result should
+be recorded based on the value of `idx`.
+
+### Handling Insert Results
+
+Unlike the other result types, `InsertOneResult` contains an `insertedId` field that is generated
+driver-side, either by recording the `_id` field present in the user's insert document or creating
+and adding one. Drivers MUST only record these `insertedId`s in a `BulkWriteResult` when a
+successful response for the insert operation (i.e. `{ "ok": 1, "n": 1 }`) is received in the
+results cursor. This ensures that drivers only report an `insertedId` when it is confirmed that the
+insert succeeded.
+
+## Handling Errors
+
+### Top-Level Errors
+
+A top-level error is any error that occurs that is not the result of a single write operation
+failing or a write concern error. Examples include network errors that occur when communicating
+with the server, command errors (`{ "ok": 0 }`) returned from the server, client-side errors, and
+errors that occur when attempting to perform a `getMore` to retrieve results from the server.
+
+When a top-level error is encountered and individual results and/or errors have already been
+observed, drivers MUST embed the top-level error within a `BulkWriteException` as the `error` field
+to retain this information. Otherwise, drivers MAY throw an exception containing only the top-level
+error.
+
+Encountering a top-level error MUST halt execution of a bulk write for both ordered and unordered
+bulk writes. This means that drivers MUST NOT attempt to retrieve more responses from the cursor or
+execute any further `bulkWrite` batches and MUST immediately throw an exception.
+
+### Write Concern Errors
+
+Write concern errors are recorded in the `writeConcernErrors` field on `BulkWriteException`. When
+a write concern error is encountered, it should not terminate execution of the bulk write for
+either ordered or unordered bulk writes. However, drivers MUST throw an exception at the end of
+execution if any write concern errors were observed.
+
+### Individual Write Errors
+
+Individual write errors retrieved from the cursor are recorded in the `writeErrors` field on
+`BulkWriteException`. If an individual write error is encountered during an ordered bulk write,
+drivers MUST record the error in `writeErrors` and immediately throw the exception. Otherwise,
+drivers MUST continue to iterate the results cursor and execute any further `bulkWrite` batches.
```

Comment:
Suggest testing `writeErrors` are collected across batches for unordered bulk writes: https://github.com/mongodb/specifications/commit/cf5743812843fdbaabee099cce85309033743c20

This revealed a bug in the C driver implementation (was not incrementing `txnNumber`).